src/block_coordinate_algorithms.jl

```diff
@@ -0,0 +1,355 @@
+                    weights = progress ./sum(progress)
+                    i = findfirst(cumsum(weights) .>= rand())
+                else
+                    @warn "Unknown update_order: $(update_order)"
+                end
+                
+                
+                v = copy(x) # This is equivalent to setting the rest of d to zero
+                indices = [idx < ndim ? Colon() : i for idx=1:ndim]
+                v[indices...] = compute_extreme_point(lmo.lmos[i], gradient[indices...])
+
+                dgi = fast_dot(x[indices...], gradient[indices...]) - fast_dot(v[indices...], gradient[indices...])
+
+                if dual_gap != Inf
+                    progress[i] = copy(dual_gaps[i]) - dgi
+                end
+                
+                dual_gaps[i] = dgi
+                dual_gap = sum(dual_gaps)
+
+                d = muladd_memory_mode(memory_mode, d, x, v)
+                gamma = perform_line_search(
+                    line_search,
+                    t,
+                    (x) -> f(x),
+                    (storage, x) -> grad!(storage, x),
+                    gradient,
+                    x,
+                    d,
+                    1.0,
+                    linesearch_workspace,
+                    memory_mode,
+                )      
+                
+                x = muladd_memory_mode(memory_mode, x, gamma, d)
+            end
+        end
+
+        
+        # go easy on the memory - only compute if really needed
+        if (
+            (mod(t, print_iter) == 0 && verbose) ||
+            callback !== nothing ||
+            line_search isa Shortstep
+        )
+            infeas = sum([fast_dot(selectdim(x,ndim,i) - selectdim(x,ndim,j), selectdim(x,ndim,i) - selectdim(x,ndim,j)) for i in 1:l for j in 1:i-1])
+            primal = f(x)
+
+        end
+       
+
+        t = t + 1
+        if callback !== nothing
+            state = CallbackStateBCFW(
+                t,
+                primal,
+                primal - dual_gap,
+                dual_gap,
+                infeas,
+                tot_time,
+                x,
+                v,
+                gamma,
+                f,
+                grad!,
+                lmo,
+                gradient,
+                tt,
+            )
+            # @show state
+            if callback(state) === false
+                break
+            end
+        end
+        
+        
+    end
+    # recompute everything once for final verfication / do not record to trajectory though for now!
+    # this is important as some variants do not recompute f(x) and the dual_gap regularly but only when reporting
+    # hence the final computation.
+    tt = last
+
+    grad!(gradient, x)
+    v = cat(compute_extreme_point(lmo, tuple([selectdim(gradient, ndim, i) for i=1:l]...))... , dims=ndim)
+    infeas = sum([fast_dot(selectdim(x,ndim,i) - selectdim(x,ndim,j), selectdim(x,ndim,i) - selectdim(x,ndim,j)) for i in 1:l for j in 1:i-1])
```

Comment:
this is allocating a vector just to sum it, one can use
```
sum(g(x) for x in thing)
```
without the square brackets